cordova-plugin-outline/apple/vpn/PacketTunnelProvider.m

```diff
@@ -496,51 +486,38 @@ - (void)restartShadowsocks:(bool)configChanged {
     return;
   }
   if (configChanged || ![activeHostNetworkAddress isEqualToString:self.hostNetworkAddress]) {
-    DDLogInfo(@"Configuration or host IP address changed with the network. Restarting ss-local.");
+    DDLogInfo(@"Configuration or host IP address changed with the network. Reconnecting tunnel.");
     self.hostNetworkAddress = activeHostNetworkAddress;
-    [self.shadowsocks stop:^(ErrorCode errorCode) {
-      DDLogInfo(@"Shadowsocks stopped.");
-      self.shadowsocks.config = [self getShadowsocksNetworkConfig];
-      __weak PacketTunnelProvider *weakSelf = self;
-      [self.shadowsocks
-          startWithConnectivityChecks:true
-                           completion:^(ErrorCode errorCode) {
-                             ErrorCode clientErrorCode =
-                                 errorCode == noError || errorCode == udpRelayNotEnabled
-                                     ? noError
-                                     : errorCode;
-                             [weakSelf execAppCallbackForAction:kActionStart
-                                                      errorCode:clientErrorCode];
-                             if (clientErrorCode != noError) {
-                               DDLogWarn(@"Tearing down VPN");
-                               [self cancelTunnelWithError:
-                                         [NSError errorWithDomain:NEVPNErrorDomain
-                                                             code:NEVPNErrorConnectionFailed
-                                                         userInfo:nil]];
-                               return;
-                             }
-                             BOOL isUdpSupported = errorCode == noError;
-                             [self setConectionUdpSupport:isUdpSupported];
-                             [weakSelf.connectionStore save:self.connection];
-                           }];
-    }];
-  }
-}
-
-- (NSDictionary *)getShadowsocksNetworkConfig {
-  if (self.connection == nil) {
-    DDLogError(@"Failed to retrieve configuration from nil connection");
-    return nil;
+    long errorCode = noError;
+    NSError *err;
+    ShadowsocksCheckConnectivity(self.hostNetworkAddress, [self.connection.port intValue],
+                                 self.connection.password, self.connection.method, &errorCode,
+                                 &err);
+    __block ErrorCode clientErrorCode =
+        (errorCode == noError || errorCode == udpRelayNotEnabled) ? noError : errorCode;
+    if (err != nil) {
```

Comment:
Can you use the return value of `ShadowsocksCheckConnectivity` instead?  I think that's equivalent but it seems to be the preferred idiom.  Then you can get rid of `err` entirely, and just pass `nil`.

Reply:
I thought that it may be useful to surface the error set by `ShadowsocksCheckConnectivity`. However, `errorCode` is always set and suffcient for the UI to display the appropriate message.